funnel/models/event.py

```diff
@@ -0,0 +1,148 @@
+
+
+def make_key():
+    key = str(uuid.uuid4()).replace('-', '')
+    return ''.join(random.sample(key, len(key)))
+
+
+def make_public_key():
+    return make_key()[:8]
+
+
+def make_private_key():
+    return make_key()[:8]
+
+
+__all__ = ['Event', 'TicketType', 'EventTicketType', 'Participant', 'Attendee', 'SyncTicket']
+
+ATTENDEE_TABLE_NAME = 'attendee'
+
+
+class EventTicketType(BaseMixin, db.Model):
+    """ Join Model for Event and TicketType
+    """
+    __tablename__ = 'event_ticket_type'
+
+    id = db.Column(db.Integer, primary_key=True, unique=True)
+    event_id = db.Column(None, db.ForeignKey('event.id'), nullable=False)
+    ticket_type_id = db.Column(None, db.ForeignKey('ticket_type.id'), nullable=False)
+
+
+class Event(BaseMixin, db.Model):
+    """ A discrete event under a proposal space.
+        For instance, a space could be associated with a workshop and a two-day conference.
+        The workshop constitutes as one event and each day of the conference
+        constitutes as an independent event.
+        This is done to allow distinguishing participants based on
+        on the tickets they have, given a participant may have a ticket
+        for only the workshop or a single day of the conference.
+        An event is associated with multiple ticket types,
+        which helps make the distinction between participants
+    """
+    __tablename__ = 'event'
+
+    id = db.Column(db.Integer, primary_key=True, unique=True)
+    name = db.Column(db.Unicode(80), nullable=False)
+    proposal_space_id = db.Column(None, db.ForeignKey('proposal_space.id'), nullable=False)
+    proposal_space = db.relationship(ProposalSpace,
+        backref=db.backref('events', cascade='all, delete-orphan', lazy='dynamic'))
+    ticket_types = db.relationship("TicketType", secondary=EventTicketType.__tablename__)
+    participants = db.relationship("Participant", secondary=ATTENDEE_TABLE_NAME)
+
+
+class TicketType(BaseMixin, db.Model):
+    """ Models different types of tickets. Eg: Early Geek, Super Early Geek, Workshop A.
+        A ticket type is associated with multiple events.
+    """
+    __tablename__ = 'ticket_type'
+
+    id = db.Column(db.Integer, primary_key=True, unique=True)
+    name = db.Column(db.Unicode(80), nullable=False)
+    proposal_space_id = db.Column(None, db.ForeignKey('proposal_space.id'), nullable=False)
+    proposal_space = db.relationship(ProposalSpace,
+        backref=db.backref('ticket_types', cascade='all, delete-orphan', lazy='dynamic'))
+    events = db.relationship("Event", secondary=EventTicketType.__tablename__)
+
+
+class Participant(BaseMixin, db.Model):
+    """ Model users participating in the proposal space
+        as an attendee, speaker, volunteer, sponsor.
+    """
+    __tablename__ = 'participant'
+
+    fullname = db.Column(db.Unicode(80), nullable=False)
+    #: Unvalidated email address
+    email = db.Column(db.Unicode(80), nullable=False, unique=True)
```

Comment:
This can't be unique. If it is, you can have only one participant record for the whole database.